File: python/rental_service.py
import sys

class Branches:

    def __init__(self):
        self.branches = {}

    def add_branch(self, branch_id, branch_name, vehicles_type):
        self.branches[branch_id] = { "branch_name": branch_name, "vehicles_type": vehicles_type }

class Vehicles:

    def __init__(self):
        self.vehicles = {}

    def add_vehicle(self, vehicle_id, vehicle_type, price):
        self.vehicles[vehicle_id] = { "vehicle_type": vehicle_type, "price": price }
        return True
    
    def get_vehicle(self, vehicle_id):
        return  self.vehicles[vehicle_id]

class BranchesVehiclesMapping:

    def __init__(self):
        self.branch_vehicle_mapping = {}

    def add_vehicle_to_branch(self, branch_id, vehicle_id):
        if branch_id not in self.branch_vehicle_mapping:
            self.branch_vehicle_mapping[branch_id] = [vehicle_id]
        else:
            self.branch_vehicle_mapping[branch_id].append(vehicle_id)
    
    def is_vehicles_to_branch_available(self, branch_id):
        if branch_id not in self.branch_vehicle_mapping:
            return False
        return True

    def get_vehicles_to_branch(self, branch_id):
        return self.branch_vehicle_mapping[branch_id]

class VehicleBooking:

    def __init__(self):
        self.booked_vehicle = {}

    def add_booking(self, vehicle_id, start_time, end_time):
        self.booked_vehicle[vehicle_id] = { "start_time": start_time, "end_time": end_time }
    
    def is_vehicle_available(self, vehicle_id, start_time, end_time):
        if vehicle_id not in self.booked_vehicle:
            return True
        else:
            booking_details =  self.booked_vehicle[vehicle_id]
            if booking_details["start_time"] >= end_time  or booking_details["end_time"] <= start_time :
                return True
            return False

class RentalService(Branches, Vehicles, BranchesVehiclesMapping, VehicleBooking):

    def __init__(self):
        Branches.__init__(self)
        Vehicles.__init__(self)
        BranchesVehiclesMapping.__init__(self)
        VehicleBooking.__init__(self)

    def add_branch(self, branch_id, branch_name, vehicles_type):
        Branches.add_branch(self, branch_id, branch_name, vehicles_type)
        print("Branch added successfully")
        
    def add_vehicle(self, branch_id, vehicle_id, vehicle_type, price):
        Vehicles.add_vehicle(self, vehicle_id, vehicle_type, price)
        BranchesVehiclesMapping.add_vehicle_to_branch(self, branch_id, vehicle_id)
        print("Vehicle added successfully")

    def get_booking_price(self, price, start_time, end_time):
        return ((end_time - start_time) * price)

    def book_vehicle(self, branch_id, vehicle_type, start_time, end_time):
        if BranchesVehiclesMapping.is_vehicles_to_branch_available(self, branch_id) == True:
            vehicle_id_list = BranchesVehiclesMapping.get_vehicles_to_branch(self, branch_id)

            booking_vehicle_id = False
            vehicle_base_price = sys.maxsize
            for vehicle_id in vehicle_id_list:
                vehicle_detail = Vehicles.get_vehicle(self, vehicle_id)
                if vehicle_detail["vehicle_type"] == vehicle_type:
                    if VehicleBooking.is_vehicle_available(self, vehicle_id, start_time, end_time) == True:
                        if vehicle_base_price > vehicle_detail["price"]:
                            vehicle_base_price = vehicle_detail["price"]
                            booking_vehicle_id = vehicle_id
                      
            if booking_vehicle_id:
                VehicleBooking.add_booking(self, booking_vehicle_id, start_time, end_time)
                print(self.get_booking_price(vehicle_base_price, start_time, end_time))
            else:
                print("-1")
        else:
            print("Invalid Branch")

    def display_vehicle(self, branch_id, start_time, end_time):
        if BranchesVehiclesMapping.is_vehicles_to_branch_available(self, branch_id) == True:
            vehicle_id_list = BranchesVehiclesMapping.get_vehicles_to_branch(self, branch_id)

            available_vehicles = {}
            for vehicle_id in vehicle_id_list:
                if VehicleBooking.is_vehicle_available(self, vehicle_id, start_time, end_time) == True:
                    vehicle_detail = Vehicles.get_vehicle(self, vehicle_id)
                    available_vehicles[vehicle_id]  = vehicle_detail

            if len(available_vehicles) > 0:
                sorted_available_vehicles = sorted(available_vehicles.items(), key = lambda x: x[1]['price'])
                for vehicle in sorted_available_vehicles:
                    print(vehicle)
            else:
                print("No vehicle available")
        else:
            print("Invalid Branch")

   

File: java/RentalService.java
import java.util.ArrayList;

import booking.Booking;
import booking.VehicleBooking;
import branch.Branch;
import branch.Branches;
import vehicle.Vehicle;
import vehicle.VehicleComparator;
import vehicle.Vehicles;

public class RentalService {
    
    private Branches branches;
    private Vehicles vehicles;
    private VehicleBooking vehicleBooking;

    public RentalService() {
        this.branches = new Branches();
        this.vehicleBooking = new VehicleBooking();
        this.vehicles = new Vehicles();
    }

    public void addBranch(Branch branch) {
        this.branches.addBranch(branch);
        System.out.println(true);
    }

    public void addVehicleToBranch(String branchName, Vehicle vehicle) {
        Branch branch = this.branches.getBranch(branchName);
        this.vehicles.putVehicleById(vehicle);
        branch.addVehicle(vehicle);
        System.out.println(true);
    }

    public double getBookingPrice(double price, int start_time, int end_time){
        return ((end_time - start_time) * price);
    }

    public void bookVehicle(String branchName, String vehicleType, int startTime, int endTime){
        Branch branch = this.branches.getBranch(branchName);
        ArrayList<Vehicle> branchVehicles= null;
        String availableVechileId = null;
        Double basePrice  = Double.MAX_VALUE;
        if(branch.isVehicleAvailable()) {
            branchVehicles = branch.getVehicles();
        
            for(Vehicle vehicle: branchVehicles) {
                if(vehicle.getVehicleType().equalsIgnoreCase(vehicleType)){
                    if(vehicleBooking.isVehicleAvailable(vehicle.getId(), startTime, endTime)){
                        if(vehicle.getPrice()< basePrice) {
                            basePrice = vehicle.getPrice();
                            availableVechileId =vehicle.getId();
                        }
                    }
                }
            }
            if(availableVechileId!= null) {
                this.vehicleBooking.addBooking(availableVechileId, new Booking(startTime, endTime));
                System.out.println(this.getBookingPrice(basePrice, startTime, endTime));
            } else {
                System.out.println("-1");
            }
        }
    }

    public void displayVehicle(String branchName, int startTime, int endTime) {
        Branch branch = this.branches.getBranch(branchName);
        ArrayList<Vehicle> branchVehicles= null;
        ArrayList<Vehicle> availableVehicles= new ArrayList<>();

        if(branch.isVehicleAvailable()) {
            branchVehicles = branch.getVehicles();
            for(Vehicle vehicle: branchVehicles) {
                if(vehicleBooking.isVehicleAvailable(vehicle.getId(), startTime, endTime)){
                    availableVehicles.add(vehicle);
                }
            }
            if(availableVehicles.size()==0) {
                return;
            }
            availableVehicles.sort(new VehicleComparator());
            for(Vehicle vehicle: availableVehicles) {
                System.out.println(vehicle.getId());
            }
        } 
    }

    public static void main(String[] args) {
        RentalService rentalService = new RentalService();

        // add branch
        rentalService.addBranch(new Branch("B1", new String[]{"CAR", "BIKE", "VAN"}));
        rentalService.addBranch(new Branch("B2", new String[]{"CAR", "VAN"}));
        rentalService.addBranch(new Branch("B3", new String[]{"CAR", "BIKE"}));

        //  add vehicles branch 1
        rentalService.addVehicleToBranch("B1", new Vehicle(1500,"CAR","V1"));
        rentalService.addVehicleToBranch("B1", new Vehicle(1000,"CAR","V2"));
        rentalService.addVehicleToBranch("B1", new Vehicle(250,"BIKE","V3"));
        rentalService.addVehicleToBranch("B1", new Vehicle(300,"VAN","V4"));
        //  add vehicles branch 2
        rentalService.addVehicleToBranch("B2", new Vehicle(100,"CAR","V5"));
        rentalService.addVehicleToBranch("B2", new Vehicle(200,"CAR","V6"));
        rentalService.addVehicleToBranch("B2", new Vehicle(1000,"VAN","V7"));
        //  add vehicles branch 3
        rentalService.addVehicleToBranch("B3", new Vehicle(2500,"BIKE","V8"));
        rentalService.addVehicleToBranch("B3", new Vehicle(300,"BIKE","V9"));
        
        //display branch 1 before booking 
        rentalService.displayVehicle("B1", 1, 5);
        //display branch 2 before booking 
        rentalService.displayVehicle("B2", 1, 5);
        //display branch 3 before booking 
        rentalService.displayVehicle("B3", 1, 5);

        // book vehicle from branch 1
        rentalService.bookVehicle("B1", "CAR", 2, 4); 
        rentalService.bookVehicle("B1", "BIKE", 1, 3);
        rentalService.bookVehicle("B1", "VAN", 6, 7);

        //display after booking
        rentalService.displayVehicle("B1", 1, 5);

        // book vehicle from branch 1
        rentalService.bookVehicle("B2", "BIKE", 1, 3); 
        rentalService.bookVehicle("B2", "VAN", 1, 3);
        rentalService.bookVehicle("B2", "VAN", 4, 7);
        rentalService.bookVehicle("B2", "CAR", 2, 3);
        rentalService.bookVehicle("B2", "CAR", 4, 5);

        //display after booking
        rentalService.displayVehicle("B2", 1, 5);

        // book vehicle from branch 1
        rentalService.bookVehicle("B3", "BIKE", 1, 3); 
        rentalService.bookVehicle("B3", "VAN", 1, 3);
        
        //display after booking
        rentalService.displayVehicle("B3", 4, 5);     
    }
}
